src/sentry/monitors/system_incidents.py

```diff
@@ -166,6 +177,258 @@ def get_clock_tick_volume_metric(tick: datetime) -> float | None:
+        return cls[st.upper()]
+
+
+class AnomalyTransition(enum.Enum):
+    ABNORMALITY_STARTED = enum.auto()
+    """
+    An abnormality has been detected during normal operations. We may
+    transition into a complete system incident, or the abnormality may recover
+    to normal.
+    """
+
+    ABNORMALITY_RECOVERED = enum.auto()
+    """
+    An abnormality has recovered back to a normal status.
+    """
+
+    INCIDENT_STARTED = enum.auto()
+    """
+    A system incident has been detected based on the historic check-in volume.
+    We are no longer able to reliably know that we are receving all check-ins.
+    """
+
+    INCIDENT_RECOVERING = enum.auto()
+    """
+    An incident has begun to recover. After this transition we will either
+    re-enter the incident va INCIDENT_STARTED or fully recover via
+    INCIDENT_RECOVERED.
+    """
+
+    INCIDENT_RECOVERED = enum.auto()
+    """
+    An incident has recovered back to normal.
+    """
+
+
+@dataclass
+class DecisionResult:
+    decision: TickAnomalyDecision
+    """
+    The recorded decision made for the clock tick
+    """
+
+    transition: AnomalyTransition | None = None
+    """
+    Reflects the transition status when making a tick decision results in a
+    state transition. None if the decision has not changed.
+    """
+
+
+def update_clock_tick_decisions(tick: datetime) -> DecisionResult:
+    """
+    Given a clock tick timestamp determine based on the historic tick volume
+    metrics, and historic tick anomaly decisions, a DecisionResult.
+
+    Once there is an INCIDENT future ticks will continue to be marked as
+    INCIDENT until we begin detecting metrics below the abnormal metric
+    threshold, at which point the system will recover after
+    MONITOR_TICK_DECISION_WINDOW ticks below the abnormal threshold.
+
+    This function will update previous decisions for earlier ticks detected as
+    ABNORMAL to either NORMAL or INCIDENT.
+    """
+    # Alias TickAnomalyDecision to improve code readability
+    Decision = TickAnomalyDecision
+
+    if not options.get("crons.tick_volume_anomaly_detection"):
+        return DecisionResult(Decision.NORMAL)
+
+    redis_client = redis.redis_clusters.get(settings.SENTRY_MONITORS_REDIS_CLUSTER)
+
+    # The clock has just ticked to the next minute. Look at the previous tick
+    # and decision metrics.
+    past_ts = tick - timedelta(minutes=1)
+
+    past_window_ts_keys = [
+        _make_reference_ts(past_ts - timedelta(minutes=delta))
+        for delta in range(0, MONITOR_TICK_DECISION_WINDOW)
+    ]
+
+    # Fetch histories for metrics and the last decision together. Window
+    # timestamps are reversed so the oldest metric is last.
+    redis_keys = chain(
+        (MONITOR_TICK_METRIC.format(ts=ts) for ts in reversed(past_window_ts_keys)),
+        (MONITOR_TICK_DECISION.format(ts=ts) for ts in [past_window_ts_keys[0]]),
+    )
+
+    histories = redis_client.mget(redis_keys)
```

Comment:
If this is a redis cluster, I think `mget` just runs a bunch of calls serially. It can be better to use a pipeline and make individual calls in one batch. Maybe just confirm this via looking at the cluster client code

Reply:
Let me follow up with this later